DCache: Add __dcacheline_aligned & __dcacheline_exclusive attributes

JarmouniA opened the pull request:
New version of #95888

Add `__dcacheline_aligned` attribute which ensures data cache line alignment for variables.

Add `__dcacheline_exclusive_noinit` & `__dcacheline_exclusive_data` attributes which make variables aligned to data cache line size and having exclusive data cache line(s).

Usage:
`static uint8_t var1 __dcacheline_aligned;`
`static uint8_t var2 __dcacheline_exclusive_noinit;`
`__dcacheline_exclusive_data static uint8_t var3 = 5;`

maass-hamburg commented on include/zephyr/linker/common-ram.ld:
```diff
@@ -151,3 +151,13 @@ ITERABLE_SECTION_RAM(scmi_protocol, Z_LINK_ITERABLE_SUBALIGN)
+		. = ALIGN(CONFIG_DCACHE_LINE_SIZE);
+		PLACE_SYMBOL_HERE(__data_cache_end);
+	} GROUP_DATA_LINK_IN(RAMABLE_REGION, RAMABLE_REGION)
+#endif /* CONFIG_DCACHE_LINE_SIZE */
```
I don't like that this automatically means, that all variables with this attribute land in the data section. That means that variables, are initialized with zero, will no longer land in the bss section. That leads to increase in size in flash, as they now can't be optimized out.

JarmouniA committed:
cache: Add __dcacheline_aligned and __dcacheline_exclusive attributes

Add __dcacheline_aligned which ensures data cache line alignment.

Add __dcacheline_exclusive_... attributes which puts a static variable
in a section that contains only data cache line aligned variables. This
results in variables being aligned and having exclusive data cache line(s).

Co-authored-by: Krzysztof Chruściński
<krzysztof.chruscinski@nordicsemi.no>
Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>
Signed-off-by: Abderrahmane JARMOUNI <git@jarmouni.me>

JarmouniA committed:
cache: replace aligned(CONFIG_DCACHE_LINE_SIZE) with dcacheline_aligned

Replace __aligned(CONFIG_DCACHE_LINE_SIZE) with __dcacheline_aligned
across Zephyr

Signed-off-by: Abderrahmane JARMOUNI <git@jarmouni.me>

JarmouniA committed:
tests: arch: cache: Add test for __dcacheline_aligned/exclusive

Add tests which checks that variable tagged with __dcacheline_aligned is
correctly aligned,
and that the variables tagged with __dcacheline_exclusive
are in the proper section that gives exclusive data cache line(s) for each.

Co-authored-by: Krzysztof Chruściński
<krzysztof.chruscinski@nordicsemi.no>
Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>
Signed-off-by: Abderrahmane JARMOUNI <git@jarmouni.me>
